File: src/go/leetcode/medium/subsets.go
package medium

// res = 加完第一个 or 不加第二个
//
func subsets(nums []int) [][]int {
	var ans [][]int
	traverse(&ans, []int{}, 0, nums)
	return ans
}

func traverse(ans *[][]int, arr []int, n int, nums []int) {
	if n == len(nums) {
		*ans = append(*ans, append([]int{}, arr...))
		return
	}
	traverse(ans, arr, n+1, nums)
	arr = append(arr, nums[n])
	traverse(ans, arr, n+1, nums)
}

File: 程序员面试经典/src/动态规划/Subsets.java
package 动态规划;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * @author Zexho
 * @date 2022/2/22 3:59 PM
 */
public class Subsets {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Stack<Integer> sub = new Stack<>();
        dp(nums, sub, 0, res);
        return res;
    }

    private void dp(int[] arr, Stack<Integer> sub, int idx, List<List<Integer>> res) {
        if (idx == arr.length) {
            if (sub.isEmpty()) res.add(new ArrayList<>(sub));
            return;
        }

        // 不加
        dp(arr, sub, idx + 1, res);

        // 加
        sub.push(arr[idx]);
        dp(arr, sub, idx + 1, res);
        sub.pop();
    }
}
